packages/labs/ssr/src/lib/module-loader.ts

```diff
@@ -0,0 +1,279 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import * as path from 'path';
+import {promises as fs} from 'fs';
+import {URL} from 'url';
+import * as vm from 'vm';
+import resolve from 'resolve';
+import {builtinModules} from 'module';
+
+type PackageJSON = {main?: string; module?: string; 'jsnext:main'?: string};
+
+const builtIns = new Set(builtinModules);
+
+const specifierMatches = (specifier: string, match: string) =>
+  specifier === match || specifier.startsWith(match + '/');
+
+export interface ModuleRecord {
+  path: string;
+  module?: vm.Module;
+  imports: Array<string>;
+  evaluated: Promise<vm.Module>;
+}
+
+interface ImportResult {
+  path: string;
+  module: vm.Module;
+}
+
+/**
+ * A JavaScript module loader that utilizes the Node `vm` module
+ * (https://nodejs.org/api/vm.html).
+ *
+ * Most of the hooks implement fairly standard web-compatible module loading:
+ *  - An import specifier resolver that uses Node module resoution
+ *  - A linker that loads dependencies from the local filesystem
+ *  - A module cache keyed by resolved URL
+ *  - import.meta.url support
+ *  - Dynamic import() that functions the same as static imports
+ *
+ * There are some behaviors specific to lit-html. Mainly that imports of certain
+ * directives are redirected to Node/SSR compatible implementations.
+ */
+export class ModuleLoader {
+  private static _nextVmContextId = 0;
+
+  // TODO (justinfagnani): document the need for this
+  private readonly _vmContextId = ModuleLoader._nextVmContextId++;
+
+  private readonly _context: vm.Context;
+
+  /**
+   * TODO (justinfagnani): This is a temporary stand-in for a real graph API.
+   * We want to be able to invalidate a module and the transitive closure
+   * of its importers so that we can update the graph.
+   *
+   * The keys of the map are useful for enumering static imported modules
+   * after an entrypoint is loaded.
+   */
+  readonly cache = new Map<string, ModuleRecord>();
+
+  // TODO (justinfagnani): Allow passing a filesystem object to allow network
+  // sources, in-memory for tests, etc.
+  constructor(contextGlobal: vm.Context) {
+    this._context = vm.createContext(contextGlobal);
+  }
+
+  /**
+   * Imports a module given by `path` into a new VM context with `contextGlobal` as the
+   * global object.
+   *
+   * @param specifier
+   * @param referrer
+   * @param contextGlobal The object that will become the global, via vm.createContext
```

Comment:
`contextGlobal` parameter doesn't exist, and the other two `@param`s have no content. Just remove the jsdoc I guess.